cli/cli.js

```diff
@@ -72,38 +76,124 @@ const runYoCommand = (cmd, args, options, opts) => {
     }
 };
 
-program.version(version).usage('[command] [options]').allowUnknownOption();
-
 const blueprintCommands = loadBlueprintCommands(blueprintsPackagePath);
 const allCommands = { ...SUB_GENERATORS, ...blueprintCommands };
 
 /* create commands */
 Object.entries(allCommands).forEach(([key, opts]) => {
-    const command = program.command(`${key} ${getArgs(opts)}`, '', { isDefault: opts.default });
+    const command = program.command(`${key} ${getArgs(opts)}`, '', { isDefault: key === 'app' });
     if (opts.alias) {
         command.alias(opts.alias);
     }
+
+    (opts.options || []).forEach(opt => {
+        command.option(opt.option, opt.desc, opt.default);
+    });
+
+    if (!opts.cliOnly) {
+        const namespace = opts.blueprint ? `${packageNameToNamespace(opts.blueprint)}:${key}` : `${JHIPSTER_NS}:${key}`;
+        const generator = env.create(namespace, { options: { help: true } });
+        Object.entries(generator._options).forEach(([key, value]) => {
+            if (value.hide || key === 'help') {
+                return;
+            }
+            let cmdString = '';
+            if (value.alias) {
+                cmdString = `-${value.alias}, `;
+            }
+            cmdString = `${cmdString}--${key}`;
+            if (value.type === String) {
+                cmdString = `${cmdString} <value>`;
+            }
+            command.option(cmdString, value.description, value.default);
+        });
+    }
+
     command
         .allowUnknownOption()
         .description(opts.desc)
-        .action(args => {
-            const options = getCommandOptions(packageJson, process.argv.slice(2));
+        .action((first, second, third) => {
```

Comment:
are thses changes to support new commander version? looks a bit complicated

Reply:
More or less. We were using commander, but parsing options manually, so:
* changed to use options parsed by the commander.
* registered options based on the generator options.
* help will work for the commands.
* commander will parse known options.

Reply:
Commander 5 changed the behaviour of the default command, so instead of relying on `program.on('command:*'` we must bail if an argument is found for app or forward to import-jdl.